subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -92,10 +95,27 @@ public DirectorySnapshotter(FileHasher hasher, Interner<String> stringInterner,
         this.collector = collector;
     }
 
-    public FileSystemLocationSnapshot snapshot(String absolutePath, @Nullable SnapshottingFilter.DirectoryWalkerPredicate predicate, final AtomicBoolean hasBeenFiltered, Consumer<FileSystemLocationSnapshot> storeInVfs) {
+    /**
+     * Snapshots a directory.
+     *
+     * Follows symlinks and includes them in the returned snapshot.
+     * Snapshots of followed symlinks are marked with {@link AccessType#VIA_SYMLINK}.
+     *
+     * @param absolutePath The absolute path of the directory to snapshot.
+     * @param predicate A predicate that determines which files to include in the snapshot.
+     *                  {@code null} means to include everything.
+     * @param hasBeenFiltered Whether the returned snapshot is a filtered snapshot of the directory.
+     * @param unfilteredSnapshotConsumer If the returned snapshot is filtered by the predicate, i.e. it doesn't have all the contents of the directory,
+     *                                   then this consumer will receive all the unfiltered snapshots within the snapshot directory.
+     *                                   For example, if an element of a directory is filtered out, the consumer will receive all the non-filtered out
+     *                                   file snapshots and all the non-filtered directory snapshots in the directory.
```

Comment:
I think it would be worth mentioning here that in case there's a filter, but it matches everything, then we handle this as being unfiltered.

Reply:
It seems we have two different concepts both discussed as "filtered".

Maybe `filtered` should mean "there is a filter" and `complete`/`incomplete` should refer to whether everything was matched or not.

WDYT?

Reply:
Hmm... `filtered` for me always implied that something wasn't matched. There is no difference currently if there is no filter or the filter matches everything. I suppose this is something I need to clarify in the docs.

Reply:
My understanding stems from `FilteredFileTree`, where "filtered" means "there is a filter", not about what is being matched. I think it would be better to capture this in the type system instead of docs only.

Reply:
We can do this in a separate PR as well if it helps move forward with the rest of the work.

Reply:
The usual filtered file tree is `DirectoryFileTree` which may or may not have a filter. Maybe `filtered` is not the right word here. I also wanted to avoid `complete`, since we already have/had `CompleteSnapshot`. I'll think about this some more.

Reply:
I created https://github.com/gradle/gradle/issues/19571 for this.